Windows timeout to websocket Chrome to short

Hello,

We are using rather slow Celeron Intel nucs for our Grafana screens. Sometimes, the startup is to slow and grafana-kiosk stops with an error. It would be great if this timeout could be extended or settable.

```
C:\xxxx\GrafanaKiosk>C:\xxxx\GrafanaKiosk\grafana-kiosk.windows.amd64.exe -c C:\xxxx\GrafanaKiosk\config.yaml
2020/08/24 10:18:40 Using config from C:\temp\GrafanaKiosk\config.yaml
2020/08/24 10:18:41 AutoFit: true
2020/08/24 10:18:41 LXDEEnabled: false
2020/08/24 10:18:41 LXDEHome: /home/pi
2020/08/24 10:18:41 Mode: full
2020/08/24 10:18:41 URL: https://grafana.xxxx.xxxx/playlists/play/9?kiosk
2020/08/24 10:18:41 LoginMethod: local
2020/08/24 10:18:41 Username: xxxx
2020/08/24 10:18:41 Password: *redacted*
2020/08/24 10:18:41 IgnoreCertificateErrors: false
2020/08/24 10:18:41 IsPlayList: true
2020/08/24 10:18:41 AutoFit: true
2020/08/24 10:18:41 LXDEEnabled: false
2020/08/24 10:18:41 LXDEHome: /home/pi
2020/08/24 10:18:41 Mode: full
2020/08/24 10:18:41 URL: https://grafana.xxxx.xxxx/playlists/play/9?kiosk
2020/08/24 10:18:41 LoginMethod: local
2020/08/24 10:18:41 Username: xxxx
2020/08/24 10:18:41 Password: *redacted*
2020/08/24 10:18:41 IgnoreCertificateErrors: false
2020/08/24 10:18:41 IsPlayList: true
2020/08/24 10:18:41 DISPLAY not set, autosetting to :0.0
2020/08/24 10:18:41 DISPLAY= :0.0
2020/08/24 10:18:41 XAUTHORITY not set, autosetting
2020/08/24 10:18:41 XAUTHORITY= /.Xauthority
2020/08/24 10:18:41 method  local
2020/08/24 10:18:41 Launching local login kiosk
panic: websocket url timeout reached

goroutine 1 [running]:
github.com/grafana/grafana-kiosk/pkg/kiosk.GrafanaKioskLocal(0xc0000c3490)
        /go/src/github.com/grafana/grafana-kiosk/pkg/kiosk/local_login.go:49 +0xc24
main.main()
        /go/src/github.com/grafana/grafana-kiosk/pkg/cmd/grafana-kiosk/main.go:153 +0x35e
```

Kind regards,
Bastiaan van Haastrecht

--- a/go.mod
+++ b/go.mod
@@ -8,7 +8,7 @@ require (
 	github.com/ilyakaznacheev/cleanenv v1.5.0
 	github.com/magefile/mage v1.15.0
 	github.com/smartystreets/goconvey v1.8.1
-	google.golang.org/api v0.168.0
+	google.golang.org/api v0.169.0
 )
 
 require (
@@ -25,7 +25,7 @@ require (
 	golang.org/x/oauth2 v0.18.0 // indirect
 	golang.org/x/text v0.14.0 // indirect
 	google.golang.org/appengine v1.6.8 // indirect
-	google.golang.org/genproto/googleapis/rpc v0.0.0-20240304212257-790db918fca8 // indirect
+	google.golang.org/genproto/googleapis/rpc v0.0.0-20240308144416-29370a3891b7 // indirect
 	google.golang.org/grpc v1.62.1 // indirect
 	google.golang.org/protobuf v1.33.0 // indirect
 )

--- a/pkg/cmd/grafana-kiosk/main.go
+++ b/pkg/cmd/grafana-kiosk/main.go
@@ -190,8 +190,8 @@ func main() {
 		cfg.GoAuth.UsernameField = args.UsernameField
 		cfg.GoAuth.PasswordField = args.PasswordField
 
-		cfg.IdToken.Audience = args.Audience
-		cfg.IdToken.KeyFile = args.KeyFile
+		cfg.IDToken.Audience = args.Audience
+		cfg.IDToken.KeyFile = args.KeyFile
 
 		cfg.ApiKey.Apikey = args.Apikey
 	}

--- a/pkg/cmd/grafana-kiosk/main.go
+++ b/pkg/cmd/grafana-kiosk/main.go
@@ -28,7 +28,7 @@ type Args struct {
 	UseMFA                  bool
 	Audience                string
 	KeyFile                 string
-	Apikey                  string
+	APIKey                  string
 	LXDEHome                string
 	ConfigPath              string
 	Mode                    string
@@ -68,7 +68,7 @@ func ProcessArgs(cfg interface{}) Args {
 	flagSettings.StringVar(&processedArgs.PasswordField, "field-password", "password", "Fieldname for the password")
 	flagSettings.StringVar(&processedArgs.Audience, "audience", "", "idtoken audience")
 	flagSettings.StringVar(&processedArgs.KeyFile, "keyfile", "key.json", "idtoken json credentials")
-	flagSettings.StringVar(&processedArgs.Apikey, "apikey", "", "apikey")
+	flagSettings.StringVar(&processedArgs.APIKey, "apikey", "", "apikey")
 
 	fu := flagSettings.Usage
 	flagSettings.Usage = func() {
@@ -193,7 +193,7 @@ func main() {
 		cfg.IDToken.Audience = args.Audience
 		cfg.IDToken.KeyFile = args.KeyFile
 
-		cfg.ApiKey.Apikey = args.Apikey
+		cfg.APIKey.APIKey = args.APIKey
 	}
 
 	// make sure the url has content
@@ -232,7 +232,7 @@ func main() {
 		kiosk.GrafanaKioskIDToken(&cfg, messages)
 	case "apikey":
 		log.Printf("Launching apikey kiosk")
-		kiosk.GrafanaKioskApikey(&cfg, messages)
+		kiosk.GrafanaKioskAPIKey(&cfg, messages)
 	case "aws":
 		log.Printf("Launcing AWS SSO kiosk")
 		kiosk.GrafanaKioskAWSLogin(&cfg, messages)

--- a/pkg/cmd/grafana-kiosk/main.go
+++ b/pkg/cmd/grafana-kiosk/main.go
@@ -134,9 +134,9 @@ func summary(cfg *kiosk.Config) {
 	log.Println("IsPlayList:", cfg.Target.IsPlayList)
 	log.Println("UseMFA:", cfg.Target.UseMFA)
 	// goauth
-	log.Println("Fieldname AutoLogin:", cfg.GOAUTH.AutoLogin)
-	log.Println("Fieldname Username:", cfg.GOAUTH.UsernameField)
-	log.Println("Fieldname Password:", cfg.GOAUTH.PasswordField)
+	log.Println("Fieldname AutoLogin:", cfg.GoAuth.AutoLogin)
+	log.Println("Fieldname Username:", cfg.GoAuth.UsernameField)
+	log.Println("Fieldname Password:", cfg.GoAuth.PasswordField)
 }
 
 func main() {
@@ -186,14 +186,14 @@ func main() {
 		cfg.General.WindowPosition = args.WindowPosition
 		cfg.General.WindowSize = args.WindowSize
 		//
-		cfg.GOAUTH.AutoLogin = args.OauthAutoLogin
-		cfg.GOAUTH.UsernameField = args.UsernameField
-		cfg.GOAUTH.PasswordField = args.PasswordField
+		cfg.GoAuth.AutoLogin = args.OauthAutoLogin
+		cfg.GoAuth.UsernameField = args.UsernameField
+		cfg.GoAuth.PasswordField = args.PasswordField
 
-		cfg.IDTOKEN.Audience = args.Audience
-		cfg.IDTOKEN.KeyFile = args.KeyFile
+		cfg.IdToken.Audience = args.Audience
+		cfg.IdToken.KeyFile = args.KeyFile
 
-		cfg.APIKEY.Apikey = args.Apikey
+		cfg.ApiKey.Apikey = args.Apikey
 	}
 
 	// make sure the url has content

--- a/pkg/cmd/grafana-kiosk/main_test.go
+++ b/pkg/cmd/grafana-kiosk/main_test.go
@@ -1,16 +1,67 @@
 package main
 
 import (
+	"os"
 	"testing"
 
+	"github.com/grafana/grafana-kiosk/pkg/kiosk"
 	. "github.com/smartystreets/goconvey/convey"
 )
 
 // TestKiosk checks kiosk command.
-func TestKioskCommand(t *testing.T) {
-	Convey("Given Kiosk CMD", t, func() {
-		Convey("cmd should...", func() {
-			So(true, ShouldBeTrue)
+func TestMain(t *testing.T) {
+	Convey("Given Default Configuration", t, func() {
+		cfg := kiosk.Config{
+			BuildInfo: kiosk.BuildInfo{
+				Version: "1.0.0",
+			},
+			General: kiosk.General{
+				AutoFit:        true,
+				LXDEEnabled:    true,
+				LXDEHome:       "/home/pi",
+				Mode:           "full",
+				WindowPosition: "0,0",
+				WindowSize:     "1920,1080",
+				ScaleFactor:    "1.0",
+			},
+			Target: kiosk.Target{
+				IgnoreCertificateErrors: false,
+				IsPlayList:              false,
+				UseMFA:                  false,
+				LoginMethod:             "local",
+				Password:                "admin",
+				URL:                     "http://localhost:3000",
+				Username:                "admin",
+			},
+			GoAuth: kiosk.GoAuth{
+				AutoLogin:     false,
+				UsernameField: "user",
+				PasswordField: "password",
+			},
+			IdToken: kiosk.IdToken{
+				KeyFile:  "/tmp/key.json",
+				Audience: "clientid",
+			},
+			ApiKey: kiosk.ApiKey{
+				Apikey: "abc",
+			},
+		}
+
+		Convey("Anonymous Login", func() {
+			oldArgs := os.Args
+			defer func() { os.Args = oldArgs }()
+			os.Args = []string{"grafana-kiosk", ""}
+			result := ProcessArgs(cfg)
+			So(result.LoginMethod, ShouldEqual, "anon")
+			So(result.URL, ShouldEqual, "https://play.grafana.org")
+		})
+		Convey("Local Login", func() {
+			oldArgs := os.Args
+			defer func() { os.Args = oldArgs }()
+			os.Args = []string{"grafana-kiosk", "-login-method", "local"}
+			result := ProcessArgs(cfg)
+			So(result.LoginMethod, ShouldEqual, "local")
+			So(result.URL, ShouldEqual, "https://play.grafana.org")
 		})
 	})
 }

--- a/pkg/cmd/grafana-kiosk/main_test.go
+++ b/pkg/cmd/grafana-kiosk/main_test.go
@@ -1,10 +1,12 @@
 package main
 
 import (
+	"log"
 	"os"
 	"testing"
 
 	"github.com/grafana/grafana-kiosk/pkg/kiosk"
+	"github.com/ilyakaznacheev/cleanenv"
 	. "github.com/smartystreets/goconvey/convey"
 )
 
@@ -46,6 +48,36 @@ func TestMain(t *testing.T) {
 				APIKey: "abc",
 			},
 		}
+		Convey("General Options", func() {
+			Convey("Parameter - autofit", func() {
+				oldArgs := os.Args
+				defer func() { os.Args = oldArgs }()
+				os.Args = []string{"grafana-kiosk", ""}
+				// starts out default true
+				result := ProcessArgs(cfg)
+				So(result.AutoFit, ShouldBeTrue)
+				// flag to set it false
+				os.Args = []string{
+					"grafana-kiosk",
+					"--autofit=false",
+				}
+				result = ProcessArgs(cfg)
+				So(result.AutoFit, ShouldBeFalse)
+			})
+
+			Convey("Environment - autofit", func() {
+				oldArgs := os.Args
+				defer func() { os.Args = oldArgs }()
+				os.Args = []string{"grafana-kiosk", ""}
+				os.Setenv("KIOSK_AUTOFIT", "false")
+				cfg := kiosk.Config{}
+				if err := cleanenv.ReadEnv(&cfg); err != nil {
+					log.Println("Error reading config from environment", err)
+				}
+				So(cfg.General.AutoFit, ShouldBeFalse)
+			})
+		})
+		// end of general options
 
 		Convey("Anonymous Login", func() {
 			oldArgs := os.Args
@@ -54,6 +86,7 @@ func TestMain(t *testing.T) {
 			result := ProcessArgs(cfg)
 			So(result.LoginMethod, ShouldEqual, "anon")
 			So(result.URL, ShouldEqual, "https://play.grafana.org")
+			So(result.AutoFit, ShouldBeTrue)
 		})
 		Convey("Local Login", func() {
 			oldArgs := os.Args
@@ -62,6 +95,7 @@ func TestMain(t *testing.T) {
 			result := ProcessArgs(cfg)
 			So(result.LoginMethod, ShouldEqual, "local")
 			So(result.URL, ShouldEqual, "https://play.grafana.org")
+			So(result.AutoFit, ShouldBeTrue)
 		})
 	})
 }

--- a/pkg/cmd/grafana-kiosk/main_test.go
+++ b/pkg/cmd/grafana-kiosk/main_test.go
@@ -38,7 +38,7 @@ func TestMain(t *testing.T) {
 				UsernameField: "user",
 				PasswordField: "password",
 			},
-			IdToken: kiosk.IdToken{
+			IDToken: kiosk.IDToken{
 				KeyFile:  "/tmp/key.json",
 				Audience: "clientid",
 			},

--- a/pkg/cmd/grafana-kiosk/main_test.go
+++ b/pkg/cmd/grafana-kiosk/main_test.go
@@ -42,8 +42,8 @@ func TestMain(t *testing.T) {
 				KeyFile:  "/tmp/key.json",
 				Audience: "clientid",
 			},
-			ApiKey: kiosk.ApiKey{
-				Apikey: "abc",
+			APIKey: kiosk.APIKey{
+				APIKey: "abc",
 			},
 		}
 

--- a/pkg/kiosk/anonymous_login.go
+++ b/pkg/kiosk/anonymous_login.go
@@ -28,15 +28,16 @@ func GrafanaKioskAnonymous(cfg *Config, messages chan string) {
 	taskCtx, cancel := chromedp.NewContext(allocCtx, chromedp.WithLogf(log.Printf))
 	defer cancel()
 
-	listenChromeEvents(taskCtx, consoleAPICall|targetCrashed)
+	listenChromeEvents(cfg, taskCtx, consoleAPICall|targetCrashed)
 
 	// ensure that the browser process is started
 	if err := chromedp.Run(taskCtx); err != nil {
 		panic(err)
 	}
 
-	// Give browser time to load next page (this can be prone to failure, explore different options vs sleeping)
-	time.Sleep(2000 * time.Millisecond)
+	// Give browser time to load
+	log.Printf("Sleeping %d MS before navigating to url", cfg.General.PageLoadDelayMS)
+	time.Sleep(time.Duration(cfg.General.PageLoadDelayMS) * time.Millisecond)
 
 	var generatedURL = GenerateURL(cfg.Target.URL, cfg.General.Mode, cfg.General.AutoFit, cfg.Target.IsPlayList)
 

--- a/pkg/kiosk/anonymous_login.go
+++ b/pkg/kiosk/anonymous_login.go
@@ -28,7 +28,7 @@ func GrafanaKioskAnonymous(cfg *Config, messages chan string) {
 	taskCtx, cancel := chromedp.NewContext(allocCtx, chromedp.WithLogf(log.Printf))
 	defer cancel()
 
-	listenChromeEvents(cfg, taskCtx, consoleAPICall|targetCrashed)
+	listenChromeEvents(taskCtx, cfg, consoleAPICall|targetCrashed)
 
 	// ensure that the browser process is started
 	if err := chromedp.Run(taskCtx); err != nil {

--- a/pkg/kiosk/apikey_login.go
+++ b/pkg/kiosk/apikey_login.go
@@ -10,8 +10,8 @@ import (
 	"github.com/chromedp/chromedp"
 )
 
-// GrafanaKioskApikey creates a chrome-based kiosk using a grafana api key.
-func GrafanaKioskApikey(cfg *Config, messages chan string) {
+// GrafanaKioskAPIKey creates a chrome-based kiosk using a grafana api key.
+func GrafanaKioskAPIKey(cfg *Config, messages chan string) {
 	dir, err := os.MkdirTemp(os.TempDir(), "chromedp-kiosk")
 	if err != nil {
 		panic(err)
@@ -47,7 +47,7 @@ func GrafanaKioskApikey(cfg *Config, messages chan string) {
 		Launch chrome and look for main-view element
 	*/
 	headers := map[string]interface{}{
-		"Authorization": "Bearer " + cfg.ApiKey.Apikey,
+		"Authorization": "Bearer " + cfg.APIKey.APIKey,
 	}
 	if err := chromedp.Run(taskCtx,
 		network.Enable(),

--- a/pkg/kiosk/apikey_login.go
+++ b/pkg/kiosk/apikey_login.go
@@ -29,7 +29,7 @@ func GrafanaKioskAPIKey(cfg *Config, messages chan string) {
 	taskCtx, cancel := chromedp.NewContext(allocCtx, chromedp.WithLogf(log.Printf))
 	defer cancel()
 
-	listenChromeEvents(cfg, taskCtx, consoleAPICall|targetCrashed)
+	listenChromeEvents(taskCtx, cfg, consoleAPICall|targetCrashed)
 
 	// ensure that the browser process is started
 	if err := chromedp.Run(taskCtx); err != nil {

--- a/pkg/kiosk/apikey_login.go
+++ b/pkg/kiosk/apikey_login.go
@@ -29,15 +29,16 @@ func GrafanaKioskApikey(cfg *Config, messages chan string) {
 	taskCtx, cancel := chromedp.NewContext(allocCtx, chromedp.WithLogf(log.Printf))
 	defer cancel()
 
-	listenChromeEvents(taskCtx, consoleAPICall|targetCrashed)
+	listenChromeEvents(cfg, taskCtx, consoleAPICall|targetCrashed)
 
 	// ensure that the browser process is started
 	if err := chromedp.Run(taskCtx); err != nil {
 		panic(err)
 	}
 
-	// Give browser time to load next page (this can be prone to failure, explore different options vs sleeping)
-	time.Sleep(2000 * time.Millisecond)
+	// Give browser time to load
+	log.Printf("Sleeping %d MS before navigating to url", cfg.General.PageLoadDelayMS)
+	time.Sleep(time.Duration(cfg.General.PageLoadDelayMS) * time.Millisecond)
 
 	var generatedURL = GenerateURL(cfg.Target.URL, cfg.General.Mode, cfg.General.AutoFit, cfg.Target.IsPlayList)
 
@@ -46,7 +47,7 @@ func GrafanaKioskApikey(cfg *Config, messages chan string) {
 		Launch chrome and look for main-view element
 	*/
 	headers := map[string]interface{}{
-		"Authorization": "Bearer " + cfg.APIKEY.Apikey,
+		"Authorization": "Bearer " + cfg.ApiKey.Apikey,
 	}
 	if err := chromedp.Run(taskCtx,
 		network.Enable(),

--- a/pkg/kiosk/aws_login.go
+++ b/pkg/kiosk/aws_login.go
@@ -29,7 +29,7 @@ func GrafanaKioskAWSLogin(cfg *Config, messages chan string) {
 	taskCtx, cancel := chromedp.NewContext(allocCtx, chromedp.WithLogf(log.Printf))
 	defer cancel()
 
-	listenChromeEvents(cfg, taskCtx, targetCrashed)
+	listenChromeEvents(taskCtx, cfg, targetCrashed)
 	// Give browser time to load
 	log.Printf("Sleeping %d MS before navigating to url", cfg.General.PageLoadDelayMS)
 	time.Sleep(time.Duration(cfg.General.PageLoadDelayMS) * time.Millisecond)

--- a/pkg/kiosk/aws_login.go
+++ b/pkg/kiosk/aws_login.go
@@ -29,7 +29,10 @@ func GrafanaKioskAWSLogin(cfg *Config, messages chan string) {
 	taskCtx, cancel := chromedp.NewContext(allocCtx, chromedp.WithLogf(log.Printf))
 	defer cancel()
 
-	listenChromeEvents(taskCtx, targetCrashed)
+	listenChromeEvents(cfg, taskCtx, targetCrashed)
+	// Give browser time to load
+	log.Printf("Sleeping %d MS before navigating to url", cfg.General.PageLoadDelayMS)
+	time.Sleep(time.Duration(cfg.General.PageLoadDelayMS) * time.Millisecond)
 
 	// ensure that the browser process is started
 	if err := chromedp.Run(taskCtx); err != nil {

--- a/pkg/kiosk/config.go
+++ b/pkg/kiosk/config.go
@@ -45,8 +45,8 @@ type IDToken struct {
 }
 
 // ApiKey
-type ApiKey struct {
-	Apikey string `yaml:"apikey" env:"KIOSK_APIKEY_APIKEY" env-description:"APIKEY"`
+type APIKey struct {
+	APIKey string `yaml:"apikey" env:"KIOSK_APIKEY_APIKEY" env-description:"APIKEY"`
 }
 
 // Config configuration for backend.
@@ -56,5 +56,5 @@ type Config struct {
 	Target    Target    `yaml:"target"`
 	GoAuth    GoAuth    `yaml:"goauth"`
 	IDToken   IDToken   `yaml:"idtoken"`
-	ApiKey    ApiKey    `yaml:"apikey"`
+	APIKey    APIKey    `yaml:"apikey"`
 }

--- a/pkg/kiosk/config.go
+++ b/pkg/kiosk/config.go
@@ -44,7 +44,7 @@ type IDToken struct {
 	Audience string `yaml:"idtoken-audience" env:"KIOSK_IDTOKEN_AUDIENCE" env-description:"Audience for idtoken, tpyically your oauth client id"`
 }
 
-// ApiKey API Key for login
+// APIKey APIKey for login
 type APIKey struct {
 	APIKey string `yaml:"apikey" env:"KIOSK_APIKEY_APIKEY" env-description:"APIKEY"`
 }

--- a/pkg/kiosk/config.go
+++ b/pkg/kiosk/config.go
@@ -1,38 +1,50 @@
 package kiosk
 
+type BuildInfo struct {
+	Version string `yaml:"version,omitempty"`
+}
+type General struct {
+	AutoFit         bool   `yaml:"autofit" env:"KIOSK_AUTOFIT" env-default:"true" env-description:"fit panels to screen"`
+	DebugEnabled    bool   `yaml:"debug" env:"KIOSK_DEBUG" env-default:"false" env-description:"enables debug output"`
+	GPUEnabled      bool   `yaml:"gpu-enabled" env:"KIOSK_GPU_ENABLED" env-default:"false" env-description:"disable GPU support"`
+	LXDEEnabled     bool   `yaml:"lxde" env:"KIOSK_LXDE_ENABLED" env-default:"false" env-description:"initialize LXDE for kiosk mode"`
+	LXDEHome        string `yaml:"lxde-home" env:"KIOSK_LXDE_HOME" env-default:"/home/pi" env-description:"path to home directory of LXDE user running X Server"`
+	Mode            string `yaml:"kiosk-mode" env:"KIOSK_MODE" env-default:"full" env-description:"[full|tv|disabled]"`
+	OzonePlatform   string `yaml:"ozone-platform" env:"KIOSK_OZONE_PLATFORM" env-default:"" env-description:"Set ozone-platform option (wayland|cast|drm|wayland|x11)"`
+	PageLoadDelayMS int64  `yaml:"page-load-delay-ms" env:"KIOSK_PAGE_LOAD_DELAY_MS" env-default:"2000" env-description:"milliseconds to wait before expecting page load"`
+	ScaleFactor     string `yaml:"scale-factor" env:"KIOSK_SCALE_FACTOR" env-default:"1.0" env-description:"Scale factor, like zoom"`
+	WindowPosition  string `yaml:"window-position" env:"KIOSK_WINDOW_POSITION" env-default:"0,0" env-description:"Top Left Position of Kiosk"`
+	WindowSize      string `yaml:"window-size" env:"KIOSK_WINDOW_SIZE" env-default:"" env-description:"Size of Kiosk in pixels (width,height)"`
+}
+type Target struct {
+	IgnoreCertificateErrors bool   `yaml:"ignore-certificate-errors" env:"KIOSK_IGNORE_CERTIFICATE_ERRORS" env-description:"ignore SSL/TLS certificate errors" env-default:"false"`
+	IsPlayList              bool   `yaml:"playlist" env:"KIOSK_IS_PLAYLIST" env-default:"false" env-description:"URL is a playlist"`
+	LoginMethod             string `yaml:"login-method" env:"KIOSK_LOGIN_METHOD" env-default:"anon" env-description:"[anon|local|gcom|goauth|idtoken|apikey]"`
+	Password                string `yaml:"password" env:"KIOSK_LOGIN_PASSWORD" env-default:"guest" env-description:"password"`
+	URL                     string `yaml:"URL" env:"KIOSK_URL" env-default:"https://play.grafana.org" env-description:"URL to Grafana server"`
+	Username                string `yaml:"username" env:"KIOSK_LOGIN_USER" env-default:"guest" env-description:"username"`
+	UseMFA                  bool   `yaml:"use-mfa" env:"KIOSK_USE_MFA" env-default:"false" env-description:"MFA is enabled for given account"`
+}
+
+type GoAuth struct {
+	AutoLogin     bool   `yaml:"auto-login" env:"KIOSK_GOAUTH_AUTO_LOGIN" env-description:"[false|true]"`
+	UsernameField string `yaml:"fieldname-username" env:"KIOSK_GOAUTH_FIELD_USER" env-description:"Username html input name value"`
+	PasswordField string `yaml:"fieldname-password" env:"KIOSK_GOAUTH_FIELD_PASSWORD" env-description:"Password html input name value"`
+}
+type IdToken struct {
+	KeyFile  string `yaml:"idtoken-keyfile" env:"KIOSK_IDTOKEN_KEYFILE" env-default:"key.json" env-description:"JSON Credentials for idtoken"`
+	Audience string `yaml:"idtoken-audience" env:"KIOSK_IDTOKEN_AUDIENCE" env-description:"Audience for idtoken, tpyically your oauth client id"`
+}
+type ApiKey struct {
+	Apikey string `yaml:"apikey" env:"KIOSK_APIKEY_APIKEY" env-description:"APIKEY"`
+}
+
 // Config configuration for backend.
 type Config struct {
-	BuildInfo struct {
-		Version string `yaml:"version,omitempty"`
-	}
-	General struct {
-		AutoFit        bool   `yaml:"autofit" env:"KIOSK_AUTOFIT" env-default:"true" env-description:"fit panels to screen"`
-		LXDEEnabled    bool   `yaml:"lxde" env:"KIOSK_LXDE_ENABLED" env-default:"false" env-description:"initialize LXDE for kiosk mode"`
-		LXDEHome       string `yaml:"lxde-home" env:"KIOSK_LXDE_HOME" env-default:"/home/pi" env-description:"path to home directory of LXDE user running X Server"`
-		Mode           string `yaml:"kiosk-mode" env:"KIOSK_MODE" env-default:"full" env-description:"[full|tv|disabled]"`
-		WindowPosition string `yaml:"window-position" env:"KIOSK_WINDOW_POSITION" env-default:"0,0" env-description:"Top Left Position of Kiosk"`
-		WindowSize     string `yaml:"window-size" env:"KIOSK_WINDOW_SIZE" env-default:"" env-description:"Size of Kiosk in pixels (width,height)"`
-		ScaleFactor    string `yaml:"scale-factor" env:"KIOSK_SCALE_FACTOR" env-default:"1.0" env-description:"Scale factor, like zoom"`
-	} `yaml:"general"`
-	Target struct {
-		IgnoreCertificateErrors bool   `yaml:"ignore-certificate-errors" env:"KIOSK_IGNORE_CERTIFICATE_ERRORS" env-description:"ignore SSL/TLS certificate errors" env-default:"false"`
-		IsPlayList              bool   `yaml:"playlist" env:"KIOSK_IS_PLAYLIST" env-default:"false" env-description:"URL is a playlist"`
-		UseMFA                  bool   `yaml:"use-mfa" env:"KIOSK_USE_MFA" env-default:"false" env-description:"MFA is enabled for given account"`
-		LoginMethod             string `yaml:"login-method" env:"KIOSK_LOGIN_METHOD" env-default:"anon" env-description:"[anon|local|gcom|goauth|idtoken|apikey]"`
-		Password                string `yaml:"password" env:"KIOSK_LOGIN_PASSWORD" env-default:"guest" env-description:"password"`
-		URL                     string `yaml:"URL" env:"KIOSK_URL" env-default:"https://play.grafana.org" env-description:"URL to Grafana server"`
-		Username                string `yaml:"username" env:"KIOSK_LOGIN_USER" env-default:"guest" env-description:"username"`
-	} `yaml:"target"`
-	GOAUTH struct {
-		AutoLogin     bool   `yaml:"auto-login" env:"KIOSK_GOAUTH_AUTO_LOGIN" env-description:"[false|true]"`
-		UsernameField string `yaml:"fieldname-username" env:"KIOSK_GOAUTH_FIELD_USER" env-description:"Username html input name value"`
-		PasswordField string `yaml:"fieldname-password" env:"KIOSK_GOAUTH_FIELD_PASSWORD" env-description:"Password html input name value"`
-	} `yaml:"goauth"`
-	IDTOKEN struct {
-		KeyFile  string `yaml:"idtoken-keyfile" env:"KIOSK_IDTOKEN_KEYFILE" env-default:"key.json" env-description:"JSON Credentials for idtoken"`
-		Audience string `yaml:"idtoken-audience" env:"KIOSK_IDTOKEN_AUDIENCE" env-description:"Audience for idtoken, tpyically your oauth client id"`
-	} `yaml:"idtoken"`
-	APIKEY struct {
-		Apikey string `yaml:"apikey" env:"KIOSK_APIKEY_APIKEY" env-description:"APIKEY"`
-	} `yaml:"apikey"`
+	BuildInfo BuildInfo `yaml:"general"`
+	General   General   `yaml:"general"`
+	Target    Target    `yaml:"target"`
+	GoAuth    GoAuth    `yaml:"goauth"`
+	IdToken   IdToken   `yaml:"idtoken"`
+	ApiKey    ApiKey    `yaml:"apikey"`
 }

--- a/pkg/kiosk/config.go
+++ b/pkg/kiosk/config.go
@@ -31,7 +31,7 @@ type Target struct {
 	UseMFA                  bool   `yaml:"use-mfa" env:"KIOSK_USE_MFA" env-default:"false" env-description:"MFA is enabled for given account"`
 }
 
-// GoAuth
+// GoAuth OAuth
 type GoAuth struct {
 	AutoLogin     bool   `yaml:"auto-login" env:"KIOSK_GOAUTH_AUTO_LOGIN" env-description:"[false|true]"`
 	UsernameField string `yaml:"fieldname-username" env:"KIOSK_GOAUTH_FIELD_USER" env-description:"Username html input name value"`
@@ -44,7 +44,7 @@ type IDToken struct {
 	Audience string `yaml:"idtoken-audience" env:"KIOSK_IDTOKEN_AUDIENCE" env-description:"Audience for idtoken, tpyically your oauth client id"`
 }
 
-// ApiKey
+// ApiKey API Key for login
 type APIKey struct {
 	APIKey string `yaml:"apikey" env:"KIOSK_APIKEY_APIKEY" env-description:"APIKEY"`
 }

--- a/pkg/kiosk/config.go
+++ b/pkg/kiosk/config.go
@@ -1,8 +1,11 @@
 package kiosk
 
+// BuildInfo contains the build version
 type BuildInfo struct {
 	Version string `yaml:"version,omitempty"`
 }
+
+// General non-site specific configuations
 type General struct {
 	AutoFit         bool   `yaml:"autofit" env:"KIOSK_AUTOFIT" env-default:"true" env-description:"fit panels to screen"`
 	DebugEnabled    bool   `yaml:"debug" env:"KIOSK_DEBUG" env-default:"false" env-description:"enables debug output"`
@@ -16,6 +19,8 @@ type General struct {
 	WindowPosition  string `yaml:"window-position" env:"KIOSK_WINDOW_POSITION" env-default:"0,0" env-description:"Top Left Position of Kiosk"`
 	WindowSize      string `yaml:"window-size" env:"KIOSK_WINDOW_SIZE" env-default:"" env-description:"Size of Kiosk in pixels (width,height)"`
 }
+
+// Target the dashboard/playlist details
 type Target struct {
 	IgnoreCertificateErrors bool   `yaml:"ignore-certificate-errors" env:"KIOSK_IGNORE_CERTIFICATE_ERRORS" env-description:"ignore SSL/TLS certificate errors" env-default:"false"`
 	IsPlayList              bool   `yaml:"playlist" env:"KIOSK_IS_PLAYLIST" env-default:"false" env-description:"URL is a playlist"`
@@ -26,15 +31,20 @@ type Target struct {
 	UseMFA                  bool   `yaml:"use-mfa" env:"KIOSK_USE_MFA" env-default:"false" env-description:"MFA is enabled for given account"`
 }
 
+// GoAuth
 type GoAuth struct {
 	AutoLogin     bool   `yaml:"auto-login" env:"KIOSK_GOAUTH_AUTO_LOGIN" env-description:"[false|true]"`
 	UsernameField string `yaml:"fieldname-username" env:"KIOSK_GOAUTH_FIELD_USER" env-description:"Username html input name value"`
 	PasswordField string `yaml:"fieldname-password" env:"KIOSK_GOAUTH_FIELD_PASSWORD" env-description:"Password html input name value"`
 }
+
+// IDToken token based login
 type IDToken struct {
 	KeyFile  string `yaml:"idtoken-keyfile" env:"KIOSK_IDTOKEN_KEYFILE" env-default:"key.json" env-description:"JSON Credentials for idtoken"`
 	Audience string `yaml:"idtoken-audience" env:"KIOSK_IDTOKEN_AUDIENCE" env-description:"Audience for idtoken, tpyically your oauth client id"`
 }
+
+// ApiKey
 type ApiKey struct {
 	Apikey string `yaml:"apikey" env:"KIOSK_APIKEY_APIKEY" env-description:"APIKEY"`
 }

--- a/pkg/kiosk/config.go
+++ b/pkg/kiosk/config.go
@@ -31,7 +31,7 @@ type GoAuth struct {
 	UsernameField string `yaml:"fieldname-username" env:"KIOSK_GOAUTH_FIELD_USER" env-description:"Username html input name value"`
 	PasswordField string `yaml:"fieldname-password" env:"KIOSK_GOAUTH_FIELD_PASSWORD" env-description:"Password html input name value"`
 }
-type IdToken struct {
+type IDToken struct {
 	KeyFile  string `yaml:"idtoken-keyfile" env:"KIOSK_IDTOKEN_KEYFILE" env-default:"key.json" env-description:"JSON Credentials for idtoken"`
 	Audience string `yaml:"idtoken-audience" env:"KIOSK_IDTOKEN_AUDIENCE" env-description:"Audience for idtoken, tpyically your oauth client id"`
 }
@@ -45,6 +45,6 @@ type Config struct {
 	General   General   `yaml:"general"`
 	Target    Target    `yaml:"target"`
 	GoAuth    GoAuth    `yaml:"goauth"`
-	IdToken   IdToken   `yaml:"idtoken"`
+	IDToken   IDToken   `yaml:"idtoken"`
 	ApiKey    ApiKey    `yaml:"apikey"`
 }

--- a/pkg/kiosk/config.go
+++ b/pkg/kiosk/config.go
@@ -51,7 +51,7 @@ type APIKey struct {
 
 // Config configuration for backend.
 type Config struct {
-	BuildInfo BuildInfo `yaml:"general"`
+	BuildInfo BuildInfo `yaml:"buildinfo"`
 	General   General   `yaml:"general"`
 	Target    Target    `yaml:"target"`
 	GoAuth    GoAuth    `yaml:"goauth"`

--- a/pkg/kiosk/grafana_com_login.go
+++ b/pkg/kiosk/grafana_com_login.go
@@ -29,7 +29,7 @@ func GrafanaKioskGCOM(cfg *Config, messages chan string) {
 	taskCtx, cancel := chromedp.NewContext(allocCtx, chromedp.WithLogf(log.Printf))
 	defer cancel()
 
-	listenChromeEvents(cfg, taskCtx, targetCrashed)
+	listenChromeEvents(taskCtx, cfg, targetCrashed)
 
 	// ensure that the browser process is started
 	if err := chromedp.Run(taskCtx); err != nil {

--- a/pkg/kiosk/grafana_com_login.go
+++ b/pkg/kiosk/grafana_com_login.go
@@ -29,12 +29,15 @@ func GrafanaKioskGCOM(cfg *Config, messages chan string) {
 	taskCtx, cancel := chromedp.NewContext(allocCtx, chromedp.WithLogf(log.Printf))
 	defer cancel()
 
-	listenChromeEvents(taskCtx, targetCrashed)
+	listenChromeEvents(cfg, taskCtx, targetCrashed)
 
 	// ensure that the browser process is started
 	if err := chromedp.Run(taskCtx); err != nil {
 		panic(err)
 	}
+	// Give browser time to load
+	log.Printf("Sleeping %d MS before navigating to url", cfg.General.PageLoadDelayMS)
+	time.Sleep(time.Duration(cfg.General.PageLoadDelayMS) * time.Millisecond)
 
 	var generatedURL = GenerateURL(cfg.Target.URL, cfg.General.Mode, cfg.General.AutoFit, cfg.Target.IsPlayList)
 

--- a/pkg/kiosk/grafana_genericoauth_login.go
+++ b/pkg/kiosk/grafana_genericoauth_login.go
@@ -29,7 +29,7 @@ func GrafanaKioskGenericOauth(cfg *Config, messages chan string) {
 	taskCtx, cancel := chromedp.NewContext(allocCtx, chromedp.WithLogf(log.Printf))
 	defer cancel()
 
-	listenChromeEvents(cfg, taskCtx, targetCrashed)
+	listenChromeEvents(taskCtx, cfg, targetCrashed)
 
 	// ensure that the browser process is started
 	if err := chromedp.Run(taskCtx); err != nil {

--- a/pkg/kiosk/grafana_genericoauth_login.go
+++ b/pkg/kiosk/grafana_genericoauth_login.go
@@ -29,7 +29,7 @@ func GrafanaKioskGenericOauth(cfg *Config, messages chan string) {
 	taskCtx, cancel := chromedp.NewContext(allocCtx, chromedp.WithLogf(log.Printf))
 	defer cancel()
 
-	listenChromeEvents(taskCtx, targetCrashed)
+	listenChromeEvents(cfg, taskCtx, targetCrashed)
 
 	// ensure that the browser process is started
 	if err := chromedp.Run(taskCtx); err != nil {
@@ -46,9 +46,9 @@ func GrafanaKioskGenericOauth(cfg *Config, messages chan string) {
 	// XPATH of grafana.com for Generic OAUTH login button = //*[@href="login/grafana_com"]/i
 
 	// Click the OAUTH login button
-	log.Println("Oauth_Auto_Login enabled: ", cfg.GOAUTH.AutoLogin)
+	log.Println("Oauth_Auto_Login enabled: ", cfg.GoAuth.AutoLogin)
 
-	if cfg.GOAUTH.AutoLogin {
+	if cfg.GoAuth.AutoLogin {
 		if err := chromedp.Run(taskCtx,
 			chromedp.Navigate(generatedURL),
 		); err != nil {
@@ -64,13 +64,15 @@ func GrafanaKioskGenericOauth(cfg *Config, messages chan string) {
 		}
 	}
 
-	// Give browser time to load next page (this can be prone to failure, explore different options vs sleeping)
-	time.Sleep(2000 * time.Millisecond)
+	// Give browser time to load
+	log.Printf("Sleeping %d MS before navigating to url", cfg.General.PageLoadDelayMS)
+	time.Sleep(time.Duration(cfg.General.PageLoadDelayMS) * time.Millisecond)
+
 	// Fill out OAUTH login page
 	if err := chromedp.Run(taskCtx,
-		chromedp.WaitVisible(`//input[@name="`+cfg.GOAUTH.UsernameField+`"]`, chromedp.BySearch),
-		chromedp.SendKeys(`//input[@name="`+cfg.GOAUTH.UsernameField+`"]`, cfg.Target.Username, chromedp.BySearch),
-		chromedp.SendKeys(`//input[@name="`+cfg.GOAUTH.PasswordField+`"]`, cfg.Target.Password+kb.Enter, chromedp.BySearch),
+		chromedp.WaitVisible(`//input[@name="`+cfg.GoAuth.UsernameField+`"]`, chromedp.BySearch),
+		chromedp.SendKeys(`//input[@name="`+cfg.GoAuth.UsernameField+`"]`, cfg.Target.Username, chromedp.BySearch),
+		chromedp.SendKeys(`//input[@name="`+cfg.GoAuth.PasswordField+`"]`, cfg.Target.Password+kb.Enter, chromedp.BySearch),
 	); err != nil {
 		panic(err)
 	}

--- a/pkg/kiosk/grafana_idtoken_login.go
+++ b/pkg/kiosk/grafana_idtoken_login.go
@@ -34,7 +34,7 @@ func GrafanaKioskIDToken(cfg *Config, messages chan string) {
 	taskCtx, cancel := chromedp.NewContext(allocCtx, chromedp.WithLogf(log.Printf))
 	defer cancel()
 
-	listenChromeEvents(cfg, taskCtx, targetCrashed)
+	listenChromeEvents(taskCtx, cfg, targetCrashed)
 
 	// ensure that the browser process is started
 	if err := chromedp.Run(taskCtx); err != nil {

--- a/pkg/kiosk/grafana_idtoken_login.go
+++ b/pkg/kiosk/grafana_idtoken_login.go
@@ -49,8 +49,8 @@ func GrafanaKioskIDToken(cfg *Config, messages chan string) {
 
 	log.Println("Navigating to ", generatedURL)
 
-	log.Printf("Token is using audience %s and reading from %s", cfg.IdToken.Audience, cfg.IdToken.KeyFile)
-	tokenSource, err := idtoken.NewTokenSource(context.Background(), cfg.IdToken.Audience, idtoken.WithCredentialsFile(cfg.IdToken.KeyFile))
+	log.Printf("Token is using audience %s and reading from %s", cfg.IDToken.Audience, cfg.IDToken.KeyFile)
+	tokenSource, err := idtoken.NewTokenSource(context.Background(), cfg.IDToken.Audience, idtoken.WithCredentialsFile(cfg.IDToken.KeyFile))
 
 	if err != nil {
 		panic(err)

--- a/pkg/kiosk/grafana_idtoken_login.go
+++ b/pkg/kiosk/grafana_idtoken_login.go
@@ -2,6 +2,7 @@ package kiosk
 
 import (
 	"context"
+	"time"
 
 	"fmt"
 	"log"
@@ -33,19 +34,23 @@ func GrafanaKioskIDToken(cfg *Config, messages chan string) {
 	taskCtx, cancel := chromedp.NewContext(allocCtx, chromedp.WithLogf(log.Printf))
 	defer cancel()
 
-	listenChromeEvents(taskCtx, targetCrashed)
+	listenChromeEvents(cfg, taskCtx, targetCrashed)
 
 	// ensure that the browser process is started
 	if err := chromedp.Run(taskCtx); err != nil {
 		panic(err)
 	}
 
+	// Give browser time to load
+	log.Printf("Sleeping %d MS before navigating to url", cfg.General.PageLoadDelayMS)
+	time.Sleep(time.Duration(cfg.General.PageLoadDelayMS) * time.Millisecond)
+
 	var generatedURL = GenerateURL(cfg.Target.URL, cfg.General.Mode, cfg.General.AutoFit, cfg.Target.IsPlayList)
 
 	log.Println("Navigating to ", generatedURL)
 
-	log.Printf("Token is using audience %s and reading from %s\n", cfg.IDTOKEN.Audience, cfg.IDTOKEN.KeyFile)
-	tokenSource, err := idtoken.NewTokenSource(context.Background(), cfg.IDTOKEN.Audience, idtoken.WithCredentialsFile(cfg.IDTOKEN.KeyFile))
+	log.Printf("Token is using audience %s and reading from %s", cfg.IdToken.Audience, cfg.IdToken.KeyFile)
+	tokenSource, err := idtoken.NewTokenSource(context.Background(), cfg.IdToken.Audience, idtoken.WithCredentialsFile(cfg.IdToken.KeyFile))
 
 	if err != nil {
 		panic(err)

--- a/pkg/kiosk/listen_chrome_events.go
+++ b/pkg/kiosk/listen_chrome_events.go
@@ -16,14 +16,14 @@ const (
 	targetCrashed
 )
 
-func listenChromeEvents(taskCtx context.Context, events chromeEvents) {
+func listenChromeEvents(cfg *Config, taskCtx context.Context, events chromeEvents) {
 	chromedp.ListenTarget(taskCtx, func(ev interface{}) {
 		switch ev := ev.(type) {
 		case *runtime.EventConsoleAPICalled:
 			if events&consoleAPICall != 0 {
-				log.Printf("console.%s call:\n", ev.Type)
+				log.Printf("console.%s call:", ev.Type)
 				for _, arg := range ev.Args {
-					log.Printf("	%s - %s\n", arg.Type, arg.Value)
+					log.Printf("	%s - %s", arg.Type, arg.Value)
 				}
 			}
 		case *inspector.EventTargetCrashed:
@@ -32,6 +32,10 @@ func listenChromeEvents(taskCtx context.Context, events chromeEvents) {
 				go func() {
 					_ = chromedp.Run(taskCtx, chromedp.Reload())
 				}()
+			}
+		default:
+			if cfg.General.DebugEnabled {
+				log.Printf("Unknown Event: %+v", ev)
 			}
 		}
 	})

--- a/pkg/kiosk/listen_chrome_events.go
+++ b/pkg/kiosk/listen_chrome_events.go
@@ -16,7 +16,7 @@ const (
 	targetCrashed
 )
 
-func listenChromeEvents(cfg *Config, taskCtx context.Context, events chromeEvents) {
+func listenChromeEvents(taskCtx context.Context, cfg *Config, events chromeEvents) {
 	chromedp.ListenTarget(taskCtx, func(ev interface{}) {
 		switch ev := ev.(type) {
 		case *runtime.EventConsoleAPICalled:

--- a/pkg/kiosk/local_login.go
+++ b/pkg/kiosk/local_login.go
@@ -30,7 +30,7 @@ func GrafanaKioskLocal(cfg *Config, messages chan string) {
 	taskCtx, cancel := chromedp.NewContext(allocCtx, chromedp.WithLogf(log.Printf))
 	defer cancel()
 
-	listenChromeEvents(cfg, taskCtx, targetCrashed)
+	listenChromeEvents(taskCtx, cfg, targetCrashed)
 
 	// ensure that the browser process is started
 	if err := chromedp.Run(taskCtx); err != nil {

--- a/pkg/kiosk/local_login.go
+++ b/pkg/kiosk/local_login.go
@@ -30,7 +30,7 @@ func GrafanaKioskLocal(cfg *Config, messages chan string) {
 	taskCtx, cancel := chromedp.NewContext(allocCtx, chromedp.WithLogf(log.Printf))
 	defer cancel()
 
-	listenChromeEvents(taskCtx, targetCrashed)
+	listenChromeEvents(cfg, taskCtx, targetCrashed)
 
 	// ensure that the browser process is started
 	if err := chromedp.Run(taskCtx); err != nil {
@@ -46,10 +46,11 @@ func GrafanaKioskLocal(cfg *Config, messages chan string) {
 		name=user, type=text
 		id=inputPassword, type=password, name=password
 	*/
-	// Give browser time to load next page (this can be prone to failure, explore different options vs sleeping)
-	time.Sleep(2000 * time.Millisecond)
+	// Give browser time to load
+	log.Printf("Sleeping %d MS before navigating to url", cfg.General.PageLoadDelayMS)
+	time.Sleep(time.Duration(cfg.General.PageLoadDelayMS) * time.Millisecond)
 
-	if cfg.GOAUTH.AutoLogin {
+	if cfg.GoAuth.AutoLogin {
 		// if AutoLogin is set, get the base URL and append the local login bypass before navigating to the full url
 		startIndex := strings.Index(cfg.Target.URL, "://") + 3
 		endIndex := strings.Index(cfg.Target.URL[startIndex:], "/") + startIndex

--- a/pkg/kiosk/utils.go
+++ b/pkg/kiosk/utils.go
@@ -50,32 +50,56 @@ func generateExecutorOptions(dir string, cfg *Config) []chromedp.ExecAllocatorOp
 	buildVersion := strings.TrimPrefix(cfg.BuildInfo.Version, "v")
 	kioskVersion := fmt.Sprintf("GrafanaKiosk/%s (%s %s)", buildVersion, runtime.GOOS, runtime.GOARCH)
 	userAgent := fmt.Sprintf("Mozilla/5.0 (X11; CrOS armv7l 13597.84.0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36 %s", kioskVersion)
+
+	// See https://peter.sh/experiments/chromium-command-line-switches/
+
+	// --start-fullscreen
+	//    Specifies if the browser should start in fullscreen mode, like if the user had pressed F11 right after startup. ↪
+	// --start-maximized
+	//    Starts the browser maximized, regardless of any previous settings
+	// --disable-gpu
+	//    Disables GPU hardware acceleration. If software renderer is not in place, then the GPU process won't launch.--disable-gpu
+
 	execAllocatorOption := []chromedp.ExecAllocatorOption{
 		chromedp.NoFirstRun,
 		chromedp.NoDefaultBrowserCheck,
-		chromedp.Flag("user-agent", userAgent),
-		chromedp.Flag("noerrdialogs", true),
-		chromedp.Flag("kiosk", true),
+		chromedp.Flag("autoplay-policy", "no-user-gesture-required"),
 		chromedp.Flag("bwsi", true),
-		chromedp.Flag("incognito", true),
-		chromedp.Flag("disable-sync", true),
+		chromedp.Flag("check-for-update-interval", "31536000"),
+		chromedp.Flag("disable-features", "Translate"),
 		chromedp.Flag("disable-notifications", true),
 		chromedp.Flag("disable-overlay-scrollbar", true),
-		chromedp.Flag("disable-features", "Translate"),
-		chromedp.Flag("window-position", cfg.General.WindowPosition),
-		chromedp.Flag("check-for-update-interval", "31536000"),
+		chromedp.Flag("disable-sync", true),
 		chromedp.Flag("ignore-certificate-errors", cfg.Target.IgnoreCertificateErrors),
-		chromedp.Flag("test-type", cfg.Target.IgnoreCertificateErrors),
-		chromedp.Flag("autoplay-policy", "no-user-gesture-required"),
+		chromedp.Flag("incognito", true),
+		chromedp.Flag("kiosk", true),
+		chromedp.Flag("noerrdialogs", true),
+		chromedp.Flag("start-fullscreen", true),
+		chromedp.Flag("start-maximized", true),
+		chromedp.Flag("user-agent", userAgent),
+		chromedp.Flag("window-position", cfg.General.WindowPosition),
 		chromedp.UserDataDir(dir),
 	}
 
-	if cfg.General.WindowSize != "" {
-		execAllocatorOption = append(execAllocatorOption, chromedp.Flag("window-size", cfg.General.WindowSize))
+	if !cfg.General.GPUEnabled {
+		execAllocatorOption = append(
+			execAllocatorOption,
+			chromedp.Flag("disable-gpu", cfg.General.GPUEnabled))
 	}
-
+	if cfg.General.OzonePlatform != "" {
+		execAllocatorOption = append(
+			execAllocatorOption,
+			chromedp.Flag("ozone-platform", cfg.General.OzonePlatform))
+	}
+	if cfg.General.WindowSize != "" {
+		execAllocatorOption = append(
+			execAllocatorOption,
+			chromedp.Flag("window-size", cfg.General.WindowSize))
+	}
 	if cfg.General.ScaleFactor != "" {
-		execAllocatorOption = append(execAllocatorOption, chromedp.Flag("force-device-scale-factor", cfg.General.ScaleFactor))
+		execAllocatorOption = append(
+			execAllocatorOption,
+			chromedp.Flag("force-device-scale-factor", cfg.General.ScaleFactor))
 	}
 
 	return execAllocatorOption

--- a/testdata/config-anon.yaml
+++ b/testdata/config-anon.yaml
@@ -1,0 +1,20 @@
+---
+general:
+  kiosk-mode: full
+  autofit: true
+  lxde: true
+  lxde-home: /home/pi
+
+target:
+  login-method: anon
+  username: user
+  password: changeme
+  playlist: false
+  use-mfa: false
+  URL: https://play.grafana.org
+  ignore-certificate-errors: false
+
+goauth:
+  auto-login: false
+  fieldname-username: username
+  fieldname-password: password

--- a/testdata/config-aws.yaml
+++ b/testdata/config-aws.yaml
@@ -1,0 +1,14 @@
+---
+general:
+  kiosk-mode: full
+  autofit: true
+  lxde: true
+  lxde-home: /home/pi
+
+target:
+  login-method: aws
+  username: bkgann@gmail.com
+  password: changeme
+  playlist: false
+  URL: https://g-0469dbd3ca.grafana-workspace.us-east-1.amazonaws.com/d/tmsOtSxZk/amazon-ec2?orgId=1
+  ignore-certificate-errors: false

--- a/testdata/config-gcom.yaml
+++ b/testdata/config-gcom.yaml
@@ -1,0 +1,19 @@
+---
+general:
+  kiosk-mode: full
+  autofit: true
+  lxde: true
+  lxde-home: /home/pi
+
+target:
+  login-method: gcom
+  username: bkgann@yahoo.com
+  password: changeme
+  playlist: false
+  URL: https://bkgann4.grafana.net
+  ignore-certificate-errors: false
+
+goauth:
+  auto-login: false
+  fieldname-username: username
+  fieldname-password: password

--- a/testdata/config-local.yaml
+++ b/testdata/config-local.yaml
@@ -1,0 +1,15 @@
+---
+general:
+  kiosk-mode: full
+  autofit: true
+  chromium-disable-update: true
+  lxde: true
+  lxde-home: /home/pi
+
+target:
+  login-method: local
+  username: user1
+  password: changeme
+  playlist: false
+  URL: https://monitor.plan3systems.com/grafana/d/BfHbSM0Mk/mining-p1?orgId=1&refresh=1m
+  ignore-certificate-errors: false

